### PR TITLE
First crack at exception handling

### PR DESCRIPTION
**Pull request by galactoise**
One question that I ran into is what to do about whether we want to end conversation.  This is an existing concern in other parts of the code (where should that flag live in a normal non-exception flow?), and so I didn't want to add a boolean to the derpwizardexception object and make the problem worse.  For now, we assume any exception ends the conversation.

### CODE REVIEW

File: voice-interface/src/main/java/com/derpgroup/derpwizard/voice/exception/DerpwizardExceptionAlexaWrapper.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    
+    SsmlOutputSpeech outputSpeech = new SsmlOutputSpeech();
+    outputSpeech.setSsml(exception.getSsmlMessage());
+    
+    SimpleCard card = new SimpleCard();
+    card.setContent(exception.getMessage());
+    card.setTitle(exception.getShortFormTextMessage());
+    
+    SpeechletResponse response = new SpeechletResponse();
+    response.setCard(card);
```
Comment (davidwphillips) [inline voice-interface/src/main/java/com/derpgroup/derpwizard/voice/exception/DerpwizardExceptionAlexaWrapper.java:36]: response.setShouldEndSession(true);

File: service/src/main/java/com/derpgroup/derpwizard/resource/AlexaResource.java [hunk 2]
```diff
@@ -74,35 +85,47 @@ public AlexaResource(MainConfig config, Environment env) {
...
+        responseEnvelope.setResponse(speechletResponse);
+      }
+  
+      return responseEnvelope;
+    }catch(DerpwizardException e){
+      LOG.debug(e.getMessage());
+      return new DerpwizardExceptionAlexaWrapper(e, "1.0",mapper.convertValue(metadata, new TypeReference<Map<String,Object>>(){}));
+    }catch(Throwable t){
+      LOG.debug(t.getMessage());
+      return new DerpwizardExceptionAlexaWrapper(new DerpwizardException(t.getMessage()),"1.0", mapper.convertValue(metadata, new TypeReference<Map<String,Object>>(){}));
```
Comment (davidwphillips) [inline service/src/main/java/com/derpgroup/derpwizard/resource/AlexaResource.java:148]: If we receive an Amazon Stop or Cancel intent, and we generate an exception while handling that, it'll bounce back up to this level and return a real response (from the DerpwizardExceptionAlexaWrapper). The Echo doesn't want a response to those amazon intents, and the Echo will blow up saying "There was a problem with the requested skill".

We have control at this level, and we know what the intent was. Inside these catch statements we can choose if we want to return an empty response, or the DerpwizardExceptionAlexaWrapper.

We can't return a response to the user on a close/cancel/abort intent anyways, so we should return an empty response to them instead.

Comment (davidwphillips) [inline service/src/main/java/com/derpgroup/derpwizard/resource/AlexaResource.java:148]: I take it back. It does seem to allow you to return both a card and speech response on cancel/abort intents. I'm not sure how to spawn a CloseSessionRequest intent though for testing. I do remember getting the "There was a problem with the requested skill" messages for something related to this though

File: voice-interface/src/main/java/com/derpgroup/derpwizard/voice/exception/DerpwizardExceptionAlexaWrapper.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.amazon.speech.json.SpeechletResponseEnvelope;
+import com.amazon.speech.speechlet.SpeechletResponse;
+import com.amazon.speech.ui.SimpleCard;
+import com.amazon.speech.ui.SsmlOutputSpeech;
+
+public class DerpwizardExceptionAlexaWrapper extends SpeechletResponseEnvelope {
+  
+  
+
```
Comment (davidwphillips) [inline voice-interface/src/main/java/com/derpgroup/derpwizard/voice/exception/DerpwizardExceptionAlexaWrapper.java]: So much room for activities

File: voice-interface/src/main/java/com/derpgroup/derpwizard/voice/exception/DerpwizardExceptionAlexaWrapper.java (hunk 1 above)
Comment (davidwphillips) [inline voice-interface/src/main/java/com/derpgroup/derpwizard/voice/exception/DerpwizardExceptionAlexaWrapper.java:36]: I still think in here we should explicitly set it to end the session

**Comment by galactoise** [comment]
@davidwphillips Addressed comments and fixed merge conflicts.

File: service/src/main/java/com/derpgroup/derpwizard/resource/AlexaResource.java [hunk 4]
```diff
@@ -78,41 +89,52 @@ public AlexaResource(MainConfig config, Environment env) {
...
+        card.setTitle("Alexa + DERP Wizard");
+  
+        @SuppressWarnings("unchecked")
+        VoiceOutput<SpeechletResponse> voiceOutput = (VoiceOutput<SpeechletResponse>) VoiceMessageFactory.buildOutputMessage(builder.build(), InterfaceType.ALEXA);
+        SpeechletResponse speechletResponse = (SpeechletResponse) voiceOutput.getImplInstance();
+        speechletResponse.setCard(card);
+  
+        responseEnvelope.setResponse(speechletResponse);
+      }
+
```
Comment (davidwphillips) [inline service/src/main/java/com/derpgroup/derpwizard/resource/AlexaResource.java:141]: Did we want this bit from Quip?:
else{
      outputSpeech = null;
      card = null;
      speechletResponse.setShouldEndSession(true);
    }

File: service/src/main/java/com/derpgroup/derpwizard/resource/AlexaResource.java [hunk 5]
```diff
@@ -78,41 +89,52 @@ public AlexaResource(MainConfig config, Environment env) {
...
+      responseEnvelope.setSessionAttributes(request.getSession().getAttributes());
+  
+      // Create a VoiceOutput object with the SSML content generated by the manater
+      if ((voiceInput.getMessageType() != MessageType.END_OF_CONVERSATION) && (voiceInput.getMessageType() != MessageType.STOP)) {
+        SimpleCard card = new SimpleCard();
+        card.setContent(builder.getRawText());
+        card.setTitle("Alexa + DERP Wizard");
+  
+        @SuppressWarnings("unchecked")
+        VoiceOutput<SpeechletResponse> voiceOutput = (VoiceOutput<SpeechletResponse>) VoiceMessageFactory.buildOutputMessage(builder.build(), InterfaceType.ALEXA);
```
Comment (davidwphillips) [inline service/src/main/java/com/derpgroup/derpwizard/resource/AlexaResource.java]: VoiceOutput<SsmlOutputSpeech> voiceOutput = (VoiceOutput<SsmlOutputSpeech>) VoiceMessageFactory.buildOutputMessage(builder.build(), InterfaceType.ALEXA);

**Pull request merged**